crates/runtime/src/memory.rs

```diff
@@ -524,80 +590,44 @@ impl Memory {
             }
         }
 
-        match self {
-            Memory::Static {
-                base,
-                size,
-                memory_image: Some(image),
-                ..
-            } => {
-                // Never exceed static memory size
-                if new_byte_size > base.len() {
-                    store.memory_grow_failed(&format_err!("static memory size exceeded"));
-                    return Ok(None);
-                }
-
-                if let Err(e) = image.set_heap_limit(new_byte_size) {
-                    store.memory_grow_failed(&e);
-                    return Ok(None);
-                }
-                *size = new_byte_size;
-            }
-            Memory::Static {
-                base,
-                size,
-                make_accessible,
-                ..
-            } => {
-                let make_accessible = make_accessible
-                    .expect("make_accessible must be Some if this is not a CoW memory");
-
-                // Never exceed static memory size
-                if new_byte_size > base.len() {
-                    store.memory_grow_failed(&format_err!("static memory size exceeded"));
-                    return Ok(None);
-                }
-
-                // Operating system can fail to make memory accessible
-                if let Err(e) = make_accessible(
-                    base.as_mut_ptr().add(old_byte_size),
-                    new_byte_size - old_byte_size,
-                ) {
-                    store.memory_grow_failed(&e);
-                    return Ok(None);
-                }
-                *size = new_byte_size;
-            }
-            Memory::Dynamic(mem) => {
-                if let Err(e) = mem.grow_to(new_byte_size) {
-                    store.memory_grow_failed(&e);
-                    return Ok(None);
-                }
+        match self.0.grow_to(new_byte_size) {
+            Ok(_) => Ok(Some(old_byte_size)),
+            Err(e) => {
+                store.memory_grow_failed(&e);
+                Ok(None)
             }
         }
-        Ok(Some(old_byte_size))
     }
 
     /// Return a `VMMemoryDefinition` for exposing the memory to compiled wasm code.
     pub fn vmmemory(&mut self) -> VMMemoryDefinition {
-        match self {
-            Memory::Static { base, size, .. } => VMMemoryDefinition {
-                base: base.as_mut_ptr().cast(),
-                current_length: *size,
-            },
-            Memory::Dynamic(mem) => mem.vmmemory(),
+        self.0.vmmemory()
+    }
+
+    /// Check if the inner implementation of [`Memory`] is indeed an
+    /// `ExternalMemory`.
+    #[cfg(feature = "pooling-allocator")]
+    pub fn is_external(&self) -> bool {
+        let as_any = &self.0 as &dyn std::any::Any;
+        as_any.downcast_ref::<ExternalMemory>().is_some()
```

Comment:
I might be doing this wrong... I'm trying to get to a `&dyn Any` so I can check the downcast is of the right kind but I'm concerned that I have the wrong type. I see a [failure](https://github.com/bytecodealliance/wasmtime/runs/6218369357?check_suite_focus=true#step:5:584) building the benchmarks that I don't think should happen. Rust help needed!

Reply:
Yeah I don't think this is correct, you'll need to change the trait's `into_any` method into something like `as_any` and use that to get the `&Any` to test

Reply:
Ok, had to add `RuntimeLinearMemory::as_any` (even more pooling-allocator-specific trait functions!) but that seems to be correct. I can't use `into_any` because that takes a `Box<Self>`.